vulture_os/toolkit/api_parser/defender/defender.py

```diff
@@ -0,0 +1,174 @@
+            params.update({'ago': 'P2M', 'limit': 5})
+        else:
+            if logs_to:
+                params.update({'untilTimeUtc': logs_to.isoformat().replace('+00:00', "Z")})
+            if logs_from:
+                params.update({'sinceTimeUtc': logs_from.isoformat().replace('+00:00', "Z")})
+
+        logger.debug(f"{[__parser__]}:get_logs: params for request are {params}", extra={'frontend': str(self.frontend)})
+
+        try:
+            response = self.oauthSession.get(
+                url,
+                params=params,
+                proxies=self.proxies
+            )
+        except OAuth2Error as e:
+            logger.error(f"{[__parser__]}:get_logs: Internal authentication error wile trying to get logs: '{str(e)}'")
+            logger.debug(f"{[__parser__]}:get_logs: API parameters: {str(self)}, request parameters: {params}")
+            raise DefenderAPIError(str(e))
+
+        if response.status_code == 401:
+            logger.error(f"{[__parser__]}:get_logs: Authentication failure while trying to get logs")
+            logger.debug(f"{[__parser__]}:get_logs: API parameters: {str(self)}, request parameters: {params}")
+            return DefenderAPIError("Authentication failed")
+        elif response.status_code == 403:
+            logger.error(f"{[__parser__]}:get_logs: Authorization error, domain(s) might not be managed by the tenant administrator")
+            logger.debug(f"{[__parser__]}:get_logs: API parameters: {str(self)}, request parameters: {params}")
+            return DefenderAPIError("Unauthorized, check Defender application parameters")
+        elif response.status_code != 200:
+            logger.error(f"{[__parser__]}:get_logs: Could not get Defender logs : status code is {response.status_code} with message '{response.text}'", extra={'frontend': str(self.frontend)})
+            raise DefenderAPIError("Could not get Defender logs")
+
+        content = response.json()
+        logger.info(f"{[__parser__]}:get_logs: got {len(content)} new lines", extra={'frontend': str(self.frontend)})
+        return content
+
+
+    def test(self):
+        try:
+            logs = self.get_logs(test=True)
+
+            return {
+                "status": True,
+                "data": _('Success')
+            }
+        except Exception as e:
+            logger.exception(f"{[__parser__]}:test: {str(e)}", extra={'frontend': str(self.frontend)})
+            return {
+                "status": False,
+                "error": str(e)
+            }
+
+
+    def execute(self):
+        # Aware UTC datetime
+        current_time = datetime.now(timezone.utc)
+        logs = self.get_logs(logs_from=self.last_api_call, logs_to=current_time)
+
+        # Downloading may take some while, so refresh token in Redis
+        self.update_lock()
+
+        if len(logs) > 0:
+
+            # All alerts are retrieved, reset timezone to now
+            self.frontend.last_api_call = current_time
```

Comment:
Make sure no log is duplicated.
Previous call last line == Current call first line.

Reply:
resolved, thanks for the catch